pkg/resource/managed_reconciler.go

```diff
@@ -323,6 +296,36 @@ func (r *ManagedReconciler) Reconcile(req reconcile.Request) (reconcile.Result,
 		return reconcile.Result{RequeueAfter: r.shortWait}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
 	}
 
+	if meta.WasDeleted(managed) {
+		// TODO: Reclaim Policy should be used between Claim and Managed. For Managed and External Resource,
+		// we need another field.
+		if observation.ResourceExists && managed.GetReclaimPolicy() == v1alpha1.ReclaimDelete {
```

Comment:
Yes, we do require that the `ExternalClient` calls need to be idempotent. `Delete` call should not return error if `deletion in progress`. Actually, `Create` call should also return OK if it gets `AlreadyExists` response from the cloud provider API. Otherwise, I think, we'd need a lock mechanism for making the calls in the perfect timing.

With the abstraction in this PR, we eliminate the need of organizing the calls for the developer. But I'm not sure if we should aim to eliminate the need that their functions need to be idempotent.

Reply:
> Yes, we do require that the ExternalClient calls to be idempotent.

That seems like a reasonable requirement. We should make sure it's documented in the Godoc for the ExternalClient interface, so that implementors know their methods must be idempotent.